libs/cardano-data/src/Data/ListMap.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Data.ListMap
+  ( ListMap (..),
```

Comment:
It might be a good idea to hide the constructor and implement `fromList` so it works similarly to `Map` wrt lookups. Right now I think the constructor adds the list elements in the opposite order to `Map.toList`.